File: src/AIModelManager.swift
import Foundation
import SwiftUI

/// AI Model Manager for integrating Qwen2.5-Coder model with MacDBG
@MainActor
public class AIModelManager: ObservableObject {
    
    // MARK: - Published Properties
    @Published public var isModelLoaded = false
    @Published public var modelName = "No Model"
    @Published public var isLoading = false
    @Published public var lastError: String?
    @Published public var aiSuggestions: [AISuggestion] = []
    
    // MARK: - Private Properties
    private let aiBridge: AIModelBridge
    private let modelPath: String
    
    // MARK: - Initialization
    public init() {
        self.aiBridge = AIModelBridge()
        
        // Get the model path from the bundle
        if let bundlePath = Bundle.main.resourcePath {
            self.modelPath = "\(bundlePath)/models/qwen2.5-coder-3b-instruct-q4_0.gguf"
        } else {
            self.modelPath = ""
        }
        
        // Try to load the actual model on initialization
        Task {
            await loadModel()
        }
    }
    
    // MARK: - Model Management
    public func loadModel() async {
        guard !modelPath.isEmpty else {
            lastError = "Model path not found"
            macdbgLog("❌ AI Model path not found in bundle", category: .error)
            return
        }
        
        // Check if model file exists
        guard FileManager.default.fileExists(atPath: modelPath) else {
            lastError = "Model file not found at path: \(modelPath)"
            macdbgLog("❌ AI Model file not found at path: \(modelPath)", category: .error)
            return
        }
        
        macdbgLog("🤖 Starting AI model load from: \(modelPath)", category: .system)
        isLoading = true
        lastError = nil
        
        // Load model on background thread
        await withCheckedContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                guard let self = self else {
                    continuation.resume()
                    return
                }
                
                macdbgLog("🤖 Calling aiBridge.loadModel...", category: .system)
                let success = self.aiBridge.loadModel(self.modelPath)
                macdbgLog("🤖 aiBridge.loadModel returned: \(success)", category: .system)
                
                DispatchQueue.main.async {
                    self.isModelLoaded = success
                    self.isLoading = false
                    
                    if success {
                        self.modelName = self.aiBridge.getModelName()
                        macdbgLog("✅ AI Model loaded successfully: \(self.modelName)", category: .system)
                    } else {
                        self.lastError = "Failed to load AI model from C++ bridge"
                        macdbgLog("❌ Failed to load AI model from C++ bridge", category: .error)
                    }
                    
                    continuation.resume()
                }
            }
        }
    }
    
    public func unloadModel() {
        aiBridge.unloadModel()
        isModelLoaded = false
        modelName = "No Model"
        aiSuggestions.removeAll()
        macdbgLog("🤖 AI Model unloaded", category: .system)
    }
    
    // MARK: - Text Generation
    public func generateText(prompt: String, maxTokens: Int = 512) async -> String {
        guard isModelLoaded else {
            let error = lastError ?? "AI model not loaded"
            macdbgLog("❌ AI generation failed: \(error)", category: .error)
            return "Error: \(error)\n\nTry restarting MacDBG to reload the AI model."
        }
        
        macdbgLog("🤖 Generating AI response for prompt: \(prompt.prefix(50))...", category: .system)
        
        return await withCheckedContinuation { continuation in
            aiBridge.generateTextAsync(prompt, maxTokens: Int32(maxTokens)) { result in
                if let result = result, !result.isEmpty {
                    macdbgLog("✅ AI generation completed successfully", category: .system)
                    continuation.resume(returning: result)
                } else {
                    macdbgLog("❌ AI generation returned empty result", category: .error)
                    continuation.resume(returning: "Error: AI model returned empty response")
                }
            }
        }
    }
    
    // MARK: - Debugger-Specific AI Features
    public func analyzeDisassembly(_ disassembly: [DisassemblyLine], context: String = "") async -> AISuggestion {
        let disassemblyText = disassembly.map { line in
            "0x\(String(format: "%llx", line.address)): \(line.instruction) \(line.operands)"
        }.joined(separator: "\n")
        
        let prompt = """
        Analyze this x64 assembly code and provide insights:
        
        Context: \(context)
        
        Assembly:
        \(disassemblyText)
        
        Please provide:
        1. What this code does
        2. Potential vulnerabilities
        3. Optimization suggestions
        4. Register usage analysis
        5. Control flow analysis
        """
        
        let analysis = await generateText(prompt: prompt, maxTokens: 1024)
        
        return AISuggestion(
            id: UUID(),
            type: .codeAnalysis,
            title: "Assembly Analysis",
            content: analysis,
            codeSnippet: disassemblyText,
            timestamp: Date(),
            address: disassembly.first?.address
        )
    }
    
    public func generateInstructionComment(_ instruction: DisassemblyLine, context: String = "") async -> AISuggestion {
        let instructionText = "0x\(String(format: "%llx", instruction.address)): \(instruction.instruction) \(instruction.operands)"
        
        let prompt = """
        Add a detailed comment for this assembly instruction:
        
        Context: \(context)
        
        Instruction: \(instructionText)
        
        Provide a clear, technical comment explaining what this instruction does and its purpose in the program flow.
        """
        
        let comment = await generateText(prompt: prompt, maxTokens: 256)
        
        return AISuggestion(
            id: UUID(),
            type: .instructionComment,
            title: "Instruction Comment",
            content: comment,
            codeSnippet: instructionText,
            timestamp: Date(),
            address: instruction.address
        )
    }
    
    public func suggestBreakpoints(_ disassembly: [DisassemblyLine], context: String = "") async -> AISuggestion {
        let disassemblyText = disassembly.map { line in
            "0x\(String(format: "%llx", line.address)): \(line.instruction) \(line.operands)"
        }.joined(separator: "\n")
        
        let prompt = """
        Suggest optimal breakpoint locations for debugging this code:
        
        Context: \(context)
        
        Code:
        \(disassemblyText)
        
        Suggest specific addresses and explain why each breakpoint would be useful for debugging.
        """
        
        let suggestions = await generateText(prompt: prompt, maxTokens: 512)
        
        return AISuggestion(
            id: UUID(),
            type: .breakpointSuggestion,
            title: "Breakpoint Suggestions",
            content: suggestions,
            codeSnippet: disassemblyText,
            timestamp: Date(),
            address: nil
        )
    }
    
    public func explainRegisterState(_ registers: [String: UInt64], context: String = "") async -> AISuggestion {
        let registerText = registers.map { (name, value) in
            "\(name): 0x\(String(format: "%llx", value))"
        }.joined(separator: "\n")
        
        let prompt = """
        Analyze this CPU register state and explain what it tells us:
        
        Context: \(context)
        
        Registers:
        \(registerText)
        
        Explain:
        1. Current execution state
        2. Function call context
        3. Data being processed
        4. Potential issues or interesting patterns
        """
        
        let explanation = await generateText(prompt: prompt, maxTokens: 512)
        
        return AISuggestion(
            id: UUID(),
            type: .registerExplanation,
            title: "Register State Analysis",
            content: explanation,
            codeSnippet: registerText,
            timestamp: Date(),
            address: nil
        )
    }
    
    public func analyzeMemoryPattern(_ memory: [UInt64: [UInt8]], context: String = "") async -> AISuggestion {
        let memoryText = memory.prefix(10).map { (address, bytes) in
            let hexBytes = bytes.map { String(format: "%02x", $0) }.joined(separator: " ")
            return "0x\(String(format: "%llx", address)): \(hexBytes)"
        }.joined(separator: "\n")
        
        let prompt = """
        Analyze this memory dump and identify patterns:
        
        Context: \(context)
        
        Memory:
        \(memoryText)
        
        Identify:
        1. Data structures
        2. Strings or text
        3. Code patterns
        4. Potential vulnerabilities
        5. Interesting values
        """
        
        let analysis = await generateText(prompt: prompt, maxTokens: 512)
        
        return AISuggestion(
            id: UUID(),
            type: .memoryAnalysis,
            title: "Memory Pattern Analysis",
            content: analysis,
            codeSnippet: memoryText,
            timestamp: Date(),
            address: nil
        )
    }
    
    // MARK: - Suggestion Management
    public func addSuggestion(_ suggestion: AISuggestion) {
        aiSuggestions.append(suggestion)
        
        // Keep only the last 50 suggestions
        if aiSuggestions.count > 50 {
            aiSuggestions.removeFirst(aiSuggestions.count - 50)
        }
    }
    
    public func clearSuggestions() {
        aiSuggestions.removeAll()
    }
    
    // MARK: - Chat Functions
    
    public func askQuestion(_ question: String) async -> AISuggestion {
        macdbgLog("🤖 Processing question: \(question)", category: .system)
        let response = await generateText(prompt: question)
        macdbgLog("🤖 Generated response: \(response.prefix(100))...", category: .system)
        
        return AISuggestion(
            id: UUID(),
            type: .aiResponse,
            title: "AI Response",
            content: response,
            timestamp: Date()
        )
    }
    
    public func analyzeCodeWithContext(code: String, question: String, context: String) async -> AISuggestion {
        let prompt = """
        Context: \(context)
        
        Code:
        \(code)
        
        Question: \(question)
        
        Please provide a detailed analysis.
        """
        
        let response = await generateText(prompt: prompt)
        
        return AISuggestion(
            id: UUID(),
            type: .codeAnalysis,
            title: "Code Analysis",
            content: response,
            timestamp: Date()
        )
    }
    
    public func removeSuggestion(_ suggestion: AISuggestion) {
        aiSuggestions.removeAll { $0.id == suggestion.id }
    }
    
    // MARK: - Configuration
    public func setTemperature(_ temperature: Float) {
        aiBridge.setTemperature(temperature)
    }
    
    public func setTopP(_ topP: Float) {
        aiBridge.setTopP(topP)
    }
    
    public func setMaxTokens(_ maxTokens: Int) {
        aiBridge.setMaxTokens(Int32(maxTokens))
    }
}

// MARK: - AI Suggestion Data Model
public struct AISuggestion: Identifiable, Codable {
    public let id: UUID
    public let type: AISuggestionType
    public let title: String
    public let content: String
    public let codeSnippet: String
    public let timestamp: Date
    public let address: UInt64?
    
    public init(id: UUID = UUID(), type: AISuggestionType, title: String, content: String, codeSnippet: String = "", timestamp: Date = Date(), address: UInt64? = nil) {
        self.id = id
        self.type = type
        self.title = title
        self.content = content
        self.codeSnippet = codeSnippet
        self.timestamp = timestamp
        self.address = address
    }
}

public enum AISuggestionType: String, CaseIterable, Codable {
    case userMessage = "User Message"
    case aiResponse = "AI Response"
    case codeAnalysis = "Code Analysis"
    case instructionComment = "Instruction Comment"
    case comment = "Comment"
    case breakpointSuggestion = "Breakpoint Suggestion"
    case registerExplanation = "Register Explanation"
    case registerAnalysis = "Register Analysis"
    case memoryAnalysis = "Memory Analysis"
    case vulnerability = "Vulnerability"
    case bugDetection = "Bug Detection"
    case optimization = "Optimization"
    case general = "General"
    
    public var icon: String {
        switch self {
        case .userMessage:
            return "person.circle.fill"
        case .aiResponse:
            return "brain.head.profile"
        case .codeAnalysis:
            return "doc.text.magnifyingglass"
        case .instructionComment:
            return "arrow.right.circle"
        case .comment:
            return "text.bubble"
        case .breakpointSuggestion:
            return "stop.circle"
        case .registerExplanation:
            return "cpu"
        case .registerAnalysis:
            return "cpu"
        case .memoryAnalysis:
            return "memorychip"
        case .vulnerability:
            return "exclamationmark.triangle"
        case .bugDetection:
            return "exclamationmark.triangle"
        case .optimization:
            return "speedometer"
        case .general:
            return "message"
        }
    }
    
    public var color: Color {
        switch self {
        case .userMessage:
            return .blue
        case .aiResponse:
            return .purple
        case .codeAnalysis:
            return .blue
        case .instructionComment:
            return .blue
        case .comment:
            return .green
        case .breakpointSuggestion:
            return .red
        case .registerExplanation:
            return .cyan
        case .registerAnalysis:
            return .orange
        case .memoryAnalysis:
            return .purple
        case .vulnerability:
            return .red
        case .bugDetection:
            return .red
        case .optimization:
            return .yellow
        case .general:
            return .primary
        }
    }
}

File: cpp/AIModelManager.cpp
#include "AIModelManager.hpp"
#include <iostream>
#include <fstream>
#include <sstream>
#include <thread>
#include <mutex>
#include <algorithm>
#include <cstring>

// Real llama.cpp integration via external llama-cli binary

AIModelManager::AIModelManager() 
    : modelLoaded_(false)
    , temperature_(0.7f)
    , topP_(0.9f)
    , maxTokens_(512)
    , contextLength_(2048)
{
    initializeLlama();
}

AIModelManager::~AIModelManager() {
    cleanupLlama();
}

bool AIModelManager::initializeLlama() {
    // Check if llama-cli is available
    int result = system("/usr/local/Cellar/llama.cpp/6390/bin/llama-cli --help >/dev/null 2>&1");
    if (result != 0) {
        std::cerr << "AI Model Manager: llama-cli not found or not executable" << std::endl;
        return false;
    }
    
    std::cout << "AI Model Manager: llama-cli backend ready for real text generation" << std::endl;
    return true;
}

void AIModelManager::cleanupLlama() {
    unloadModel();
    // No cleanup needed for external llama-cli
}

bool AIModelManager::loadModel(const std::string& modelPath) {
    std::lock_guard<std::mutex> lock(modelMutex_);
    
    if (modelLoaded_) {
        unloadModel();
    }
    
    // Check if model file exists
    std::ifstream file(modelPath);
    if (!file.good()) {
        std::cerr << "AI Model Manager: Model file not found: " << modelPath << std::endl;
        return false;
    }
    
    modelPath_ = modelPath;
    
    // Extract model name from path
    size_t lastSlash = modelPath.find_last_of("/\\");
    if (lastSlash != std::string::npos) {
        modelName_ = modelPath.substr(lastSlash + 1);
    } else {
        modelName_ = modelPath;
    }
    
    std::cout << "AI Model Manager: Preparing GGUF model: " << modelName_ << std::endl;
    std::cout << "AI Model Manager: Model path: " << modelPath << std::endl;
    
    // Test that llama-cli can access the model
    std::ostringstream test_cmd;
    test_cmd << "/usr/local/Cellar/llama.cpp/6390/bin/llama-cli";
    test_cmd << " -m \"" << modelPath << "\"";
    test_cmd << " -p \"test\"";
    test_cmd << " -n 1";
    test_cmd << " 2>/dev/null >/dev/null"; // Suppress all output
    
    int test_result = system(test_cmd.str().c_str());
    if (test_result != 0) {
        std::cerr << "Failed to validate GGUF model with llama-cli" << std::endl;
        return false;
    }
    
    modelLoaded_ = true;
    std::cout << "AI Model Manager: GGUF model validated and ready for real text generation!" << std::endl;
    
    return true;
}

void AIModelManager::unloadModel() {
    std::lock_guard<std::mutex> lock(modelMutex_);
    
    if (!modelLoaded_) {
        return;
    }
    
    // No complex cleanup needed since we use llama-cli externally
    modelLoaded_ = false;
    modelPath_.clear();
    modelName_.clear();
    
    std::cout << "AI Model Manager: GGUF model unloaded" << std::endl;
}

bool AIModelManager::isModelLoaded() const {
    std::lock_guard<std::mutex> lock(modelMutex_);
    return modelLoaded_;
}

std::string AIModelManager::generateText(const std::string& prompt, int maxTokens) {
    std::lock_guard<std::mutex> lock(modelMutex_);
    
    if (!modelLoaded_) {
        return "Error: No model loaded";
    }
    
    std::cout << "AI Model Manager: Generating real text using llama-cli for prompt: " << prompt.substr(0, 50) << "..." << std::endl;
    
    // Use llama-cli binary to generate text with the loaded model
    // This is simpler and more reliable than implementing the complex llama.cpp API
    
    // Create a temporary file for the prompt
    std::string temp_prompt_file = "/tmp/macdbg_prompt.txt";
    std::ofstream prompt_file(temp_prompt_file);
    if (!prompt_file.is_open()) {
        return "Error: Could not create temporary prompt file";
    }
    prompt_file << prompt;
    prompt_file.close();
    
    // Construct the llama-cli command
    std::ostringstream cmd;
    cmd << "/usr/local/Cellar/llama.cpp/6390/bin/llama-cli";
    cmd << " -m \"" << modelPath_ << "\"";
    cmd << " -p \"" << prompt << "\"";
    cmd << " -n " << maxTokens;
    cmd << " --temp 0.7";
    cmd << " --top-p 0.9";
    cmd << " --repeat-penalty 1.1";
    cmd << " --ctx-size 2048";
    cmd << " 2>/dev/null"; // Suppress stderr
    
    std::cout << "AI Model Manager: Running command: " << cmd.str().substr(0, 100) << "..." << std::endl;
    
    // Execute the command and capture output
    FILE* pipe = popen(cmd.str().c_str(), "r");
    if (!pipe) {
        return "Error: Failed to execute llama-cli";
    }
    
    std::string response;
    char buffer[4096];
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        response += buffer;
    }
    
    int exit_code = pclose(pipe);
    
    // Clean up temporary file
    std::remove(temp_prompt_file.c_str());
    
    if (exit_code != 0) {
        return "Error: llama-cli execution failed with code " + std::to_string(exit_code);
    }
    
    // Remove the prompt echo from the response (llama-cli often echoes the prompt)
    size_t prompt_end = response.find(prompt);
    if (prompt_end != std::string::npos) {
        response = response.substr(prompt_end + prompt.length());
    }
    
    // Remove common llama-cli artifacts and control text
    std::vector<std::string> artifacts = {
        "EOF by user",
        "> EOF by user",
        "assistant ",
        "user:",
        "assistant:",
        "> ",
        "User:",
        "Assistant:",
        "\n\n> ",
        "\n> ",
        ">EOF",
        "EOF"
    };
    
    for (const auto& artifact : artifacts) {
        size_t pos = 0;
        while ((pos = response.find(artifact, pos)) != std::string::npos) {
            response.erase(pos, artifact.length());
        }
    }
    
    // Remove lines that start with ">" (common in llama-cli output)
    std::istringstream iss(response);
    std::ostringstream cleaned;
    std::string line;
    
    while (std::getline(iss, line)) {
        // Skip lines that start with ">" or are empty/whitespace
        if (!line.empty() && line[0] != '>' && line.find_first_not_of(" \t\n\r") != std::string::npos) {
            cleaned << line << "\n";
        }
    }
    
    response = cleaned.str();
    
    // Trim whitespace
    size_t start = response.find_first_not_of(" \t\n\r");
    if (start != std::string::npos) {
        response = response.substr(start);
    }
    
    size_t end = response.find_last_not_of(" \t\n\r");
    if (end != std::string::npos) {
        response = response.substr(0, end + 1);
    }
    
    std::cout << "AI Model Manager: Generated real response: " << response.substr(0, 100) << "..." << std::endl;
    return response.empty() ? "Error: No response generated" : response;
}

std::string AIModelManager::generateCodeAnalysis(const std::string& disassembly, const std::string& context) {
    std::string prompt = "Analyze this assembly code and provide insights:\n\n";
    prompt += "Context: " + context + "\n\n";
    prompt += "Assembly:\n" + disassembly + "\n\n";
    prompt += "Please provide:\n";
    prompt += "1. What this code does\n";
    prompt += "2. Potential vulnerabilities\n";
    prompt += "3. Optimization suggestions\n";
    prompt += "4. Register usage analysis\n";
    
    return generateText(prompt, 1024);
}

std::string AIModelManager::generateComment(const std::string& instruction, const std::string& context) {
    std::string prompt = "Add a detailed comment for this assembly instruction:\n\n";
    prompt += "Context: " + context + "\n\n";
    prompt += "Instruction: " + instruction + "\n\n";
    prompt += "Provide a clear, technical comment explaining what this instruction does:";
    
    return generateText(prompt, 256);
}

std::string AIModelManager::generateBreakpointSuggestion(const std::string& code, const std::string& context) {
    std::string prompt = "Suggest optimal breakpoint locations for debugging this code:\n\n";
    prompt += "Context: " + context + "\n\n";
    prompt += "Code:\n" + code + "\n\n";
    prompt += "Suggest specific addresses and explain why each breakpoint would be useful:";
    
    return generateText(prompt, 512);
}

void AIModelManager::generateTextAsync(const std::string& prompt, 
                                     std::function<void(const std::string&)> completion,
                                     int maxTokens) {
    // Run generation in background thread
    std::thread([this, prompt, completion, maxTokens]() {
        std::string result = generateText(prompt, maxTokens);
        completion(result);
    }).detach();
}

void AIModelManager::setTemperature(float temperature) {
    std::lock_guard<std::mutex> lock(modelMutex_);
    temperature_ = std::max(0.0f, std::min(2.0f, temperature));
}

void AIModelManager::setTopP(float topP) {
    std::lock_guard<std::mutex> lock(modelMutex_);
    topP_ = std::max(0.0f, std::min(1.0f, topP));
}

void AIModelManager::setMaxTokens(int maxTokens) {
    std::lock_guard<std::mutex> lock(modelMutex_);
    maxTokens_ = std::max(1, std::min(4096, maxTokens));
}

std::string AIModelManager::getModelName() const {
    std::lock_guard<std::mutex> lock(modelMutex_);
    return modelName_;
}

size_t AIModelManager::getModelSize() const {
    // TODO: Return actual model size
    return 0;
}

int AIModelManager::getContextLength() const {
    return contextLength_;
}

std::string AIModelManager::processTokens(const std::vector<int>& tokens) {
    // TODO: Implement token processing
    // This would convert token IDs back to text
    return "Token processing not implemented";
}
